### PR TITLE
add xstream

### PR DESCRIPTION
**Pull request by Sophie-Normand**
We are currently working to implement PySAD into River, for the "polytechnique-project".
It is a pull request for adding the method "xstream".

Sophie Normand

### CODE REVIEW

File: river/utils/math.py [hunk 1]
```diff
@@ -29,31 +29,56 @@
...
+	return min, max
+    
+def _iterate(X, y=None):
+    """Iterates array of features and possibly labels.
+    Args:
+        X (np.array of shape (num_instances, num_features)): The features array.
+        y (np.array of shape (num_instances, ): The array containing labels (Default=None).
+    """
+
+    iterator = ArrayStreamer(shuffle=False)
```
Comment (MaxHalford) [inline river/utils/math.py:55]: Do you use this function? I don't see where `ArrayStreamer` is imported

File: river/utils/math.py [hunk 2]
```diff
@@ -29,31 +29,56 @@
...
+##################################################################################
 
 def dotvecmat(x, A):
     """Vector times matrix from left side, i.e. transpose(x)A.
-
     Parameters
     ----------
     x
     A
-
```
Comment (MaxHalford) [inline river/utils/math.py:40]: Your code editor is removing empty lines, but it shouldn't be. They increase readibility

File: river/utils/math.py [hunk 3]
```diff
@@ -336,20 +315,16 @@ def sherman_morrison(A: np.ndarray, u: np.ndarray, v: np.ndarray):
...
     References
     ----------
     [^1]: [Matrix inverse mini-batch updates — Max Halford](https://maxhalford.github.io/blog/matrix-inverse-mini-batch/)
-
     """
     eye = np.eye(len(V))
     Au = A @ U
-    A -= Au @ np.linalg.inv(eye + V @ Au) @ V @ A
+    A -= Au @ np.linalg.inv(eye + V @ Au) @ V @ A
```
Comment (MaxHalford) [inline river/utils/math.py:330]: Your code editor shouldn't remove the last empty line of a file

File: river/anomaly/base_transformer.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+#import abc
+import numpy as np
+import math
+import utils
+
+#from pysad.utils import _iterate
+
+
+class BaseTransformer(ABC):
```
Comment (MaxHalford) [inline river/anomaly/base_transformer.py:11]: What does this class do differently than the existing `Transformer` class in the `base` module? Do you really need to introduce a new base class?

File: river/anomaly/xstream.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+#from pysad.utils import get_minmax_array
+
+
+class xStream(anomaly.base.AnomalyDetector):
+    """The xStream model for row-streaming data :cite:`xstream`. It first projects the data via streamhash projection. It then fits half space chains by reference windowing. It scores the instances using the window fitted to the reference window.
+    Args:
+        n_components (int): The number of components for streamhash projection (Default=100).
+        n_chains (int): The number of half-space chains (Default=100).
+        depth (int): The maximum depth for the chains (Default=25).
+        window_size (int): The size (and the sliding length) of the reference window (Default=25).
```
Comment (MaxHalford) [inline river/anomaly/xstream.py:21]: As you can see in the rest of River's codebase, we don't use this format for documenting class parameters.

File: river/anomaly/xstream.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        # mn = np.min(np.concatenate(self.ref_window, axis=0), axis=0)
+        mn, mx = get_minmax_array(np.concatenate(self.ref_window, axis=0))
+
+        deltamax = (mx - mn) / 2.0
+        deltamax[np.abs(deltamax) <= 0.0001] = 1.0
+
+        return deltamax
+
+
+class _Chain:
```
Comment (MaxHalford) [inline river/anomaly/xstream.py:98]: Can you say what this class does? It's a bit difficult to understand from just looking at the code

**Review by MaxHalford — CHANGES_REQUESTED** [review]
Hey there! Thanks for issuing this PR, as xStream is a valuable algorithm to add.

There are quite some changes to do for this to be accepted in the codebase. I've made a few comments in a first pass. The main issue really is that you're using numpy, whereas we mostly try to avoid numpy in River. Ideally you should reimplement the logic with pure Python. 

PS: Have you installed the `pre-commit` hooks as documented in `CONTRIBUTING.md`? Those should help you catch a lot of code style issues

**Comment by MaxHalford** [comment]
@Sophie-Normand is it ok if I close this PR?

**Comment by Sophie-Normand** [comment]
> 

Yes, sure. You can close this PR

**Pull request closed without merging**